### PR TITLE
Optimisation of `mlRewrite`, `ProofInfo`

### PR DESCRIPTION
**Pull request by berpeti**
This pull request includes the following:

- Simplification of proof annotations (`FP` part is omitted).
- The ordering of proof infos are separated from the semantics of proof infos.
- `try_solve_pile` tactic has been enhanced.
- `use <info> in <hypo>` tactic changes the proof annotation used in `<hypo>` to `<info>`, and tries to solve the generated obligation with `try_solve_pile`.
- A number of theorems have been generalized, so that they can use arbitrary fresh variables instead of generated ones.

### CODE REVIEW

File: matching-logic/src/ProofSystem.v [hunk 1]
```diff
@@ -593,22 +600,75 @@ Section proof_info.
...
+    constructor. 1-2: set_solver.
+    apply implb_true_iff.
+    pose proof (proj1 (implb_true_iff _ _) HKT).
+    pose proof (proj1 (implb_true_iff _ _) pwi_pf_kt0).
+    tauto.
+  Qed.
+
+End proof_info.
+
+  Ltac convert_implb :=
```
Comment (h0nzZik) [inline matching-logic/src/ProofSystem.v:625]: I would like to see these tactic moved somewhere else (into the Utils directory? ) But we can do that in a separate PR.

File: matching-logic/src/SyntaxLemmas/PatternCtxApplicationCtx.v [hunk 2]
```diff
@@ -189,6 +188,7 @@ Proof.
   apply ApplicationContext2PatternCtx2ApplicationContext'.
 Qed.
 
+(* TODO: This needs to use count_evar_occurrences, because the prover depends on it *)
```
Comment (h0nzZik) [inline matching-logic/src/SyntaxLemmas/PatternCtxApplicationCtx.v:191]: Can you elaborate on that comment?

**Review by h0nzZik — APPROVED** [review]

File: matching-logic/src/SyntaxLemmas/PatternCtxApplicationCtx.v (hunk 2 above)
Comment (berpeti) [reply matching-logic/src/SyntaxLemmas/PatternCtxApplicationCtx.v:191]: I tried to rewrite it using `<var> \notin free_evars <pattern>`, but couldn't adjust the proofs in the prover that depend on `is_implicative_context`. (Apparently, we only use this concept in the prover currently.)
